### PR TITLE
feat: create source post moderation

### PR DESCRIPTION
**Pull request by AmarTrebinjac**
Created a new PR for creating source post moderation because [the previous one became a mess](https://github.com/dailydotdev/daily-api/pull/2373). Between the weird initial commit history and splitting off half the PR into another one and renaming a bunch of stuff, I thought it'd be easier to just make a new PR and avoid dupes and merge conflicts. Still accounting for the feedback given in the previous PR.

### CODE REVIEW

File: __tests__/posts.ts
Comment (ilasw) [inline __tests__/posts.ts:3474]: I think we should add some tests for user (member/mod/admin) trying to post in squad directly. 
This is an important point for this PR imo 👀

File: src/routes/boot.ts
Comment (ilasw) [inline src/routes/boot.ts:202]: Why are we changing this? If we are Moderator we can also post, right?

File: src/schema/posts.ts
Comment (ilasw) [inline src/schema/posts.ts:1674]: I think `id` should be autogenerated in insert, do we need to do it here?

**Review by ilasw — CHANGES_REQUESTED** [review]
Seems good code wise, mainly we should add some more tests

File: src/schema/posts.ts
Comment (AmarTrebinjac) [reply src/schema/posts.ts:1674]: This is how we do it in the other requests where we return the entity that was just created with graphql as far as I can tell

File: __tests__/posts.ts
Comment (sshanzel) [inline __tests__/posts.ts:3535]: It should not actually be possible to create a type of `article` as that only comes from the scraper.

File: __tests__/posts.ts
Comment (sshanzel) [inline __tests__/posts.ts:3523]: For `share` type, there is to types, one when there is a `sharePostId`, and one when the `externalLink` is present.

**Review by sshanzel — COMMENTED** [review]
Looks amazing overall, I think we just missing a few more tests but we are almost there 🤩

File: __tests__/posts.ts
Comment (AmarTrebinjac) [reply __tests__/posts.ts:3614]: @sshanzel Are you sure? the RFC mentions the type can be article, share or freeform. I assumed Article was a shared link. Looking at createExternalLink , it adds a post of type article:

```
 await entityManager.getRepository(ArticlePost).insert({
```

File: src/routes/boot.ts
Comment (rebelchris) [reply src/routes/boot.ts:203]: But moderator can always bypass which you already know based on currentUser when posting.
I also don't think we need this change.

File: src/schema/posts.ts
Comment (rebelchris) [inline src/schema/posts.ts:1665]: Doesn't `canAccessSource` handle this already?

File: src/schema/posts.ts
Comment (rebelchris) [reply src/schema/posts.ts:1674]: Leaving this up to @capJavert we do use both ways.
Think uuid only for article related stuff, but it's more for yggdrasil I think.

File: src/schema/posts.ts
Comment (rebelchris) [inline src/schema/posts.ts:1711]: If we going to query anyway, you might as well just get the return value from the insert?

**Review by rebelchris — COMMENTED** [review]
Very minor feedback, overall good approach.

File: __tests__/fixture/source.ts
Comment (capJavert) [inline __tests__/fixture/source.ts:89]: I would just define this in the file you are using it, we want to move away from fixtures probably since they often create flaky tests

File: src/schema/posts.ts
Comment (capJavert) [reply src/schema/posts.ts:1674]: We don't have to generate UUID, psql will automatically generate it since its a primary column.

File: src/schema/sources.ts
Comment (capJavert) [inline src/schema/sources.ts:962]: why is this needed specifically, does below case not solve it?

File: src/schema/posts.ts
Comment (capJavert) [reply src/schema/posts.ts:1665]: +1

also do we want to check type of source as well, and throw if its not a squad, just to make sure no accidental moderations for sources if someone is playing with API directly, you can see example in `joinSource`

File: src/common/post.ts
Comment (capJavert) [inline src/common/post.ts:249]: We should return the result from `save` because it will have PK and other defaults from the DB.

File: src/schema/posts.ts
Comment (AmarTrebinjac) [reply src/schema/posts.ts:1711]: We do this on other post creation as well. Asked @capJavert about it and the reason we do it this way is to retain the GraphQL schema

File: src/schema/posts.ts
Comment (AmarTrebinjac) [reply src/schema/posts.ts:1665]: @rebelchris it doesn't because source is not an optional parameter

File: src/schema/sources.ts
Comment (AmarTrebinjac) [reply src/schema/sources.ts:962]: @capJavert  We need it so moderators can bypass the moderation required check. If we remove this line, the next one will return false even if user is moderator or higher.

File: src/schema/posts.ts
Comment (rebelchris) [reply src/schema/posts.ts:1716]: Yeah aware of that, meant more you could maybe return the inserted ID, then return the graphORM layer?

File: src/routes/boot.ts
Comment (rebelchris) [reply src/routes/boot.ts:203]: @capJavert / @ilasw  I guess since we decided to handle deciding if it's moderated post on frontend, they need it in boot, as we have flows (context -> share to squad) that rely soley on boot squads now.

(Just FYI, as I struggled to understand the case at first)
Thanks @AmarTrebinjac for explaining the edge-case your facing.

Comment (capJavert) [reply src/routes/boot.ts:203]: ok from my side 👍

Comment (AmarTrebinjac) [reply src/routes/boot.ts:203]: The TL:DR is basically: 

Post flows use the squads provided by boot, and not the source query

File: src/schema/posts.ts
Comment (AmarTrebinjac) [reply src/schema/posts.ts]: @rebelchris the await is in a transaction, and the return is outside that scope. But looking on it, there's not really any need for the transaction, so I'll remove it and it should be good.

File: __tests__/posts.ts
Comment (sshanzel) [reply __tests__/posts.ts:3662]: Yes. We may create the article post behind the scenes but we don't tag it as created by the user. We will create another post with `Share` type that is then referenced to the logged user.

Comment (sshanzel) [reply __tests__/posts.ts:3662]: This means, your request is of `share` type than `article.

**Review by rebelchris — APPROVED** [review]
Not much to add from my side.
Nice work

**Review by sshanzel — APPROVED** [review]
Don't have anything blocking! Looks amazing already, just want to see the same tests be applied where applicable.

**Review by sshanzel — APPROVED** [review]
Amazing work. Once @capJavert approves, I think we should be safe to merge.

File: src/schema/sources.ts
Comment (capJavert) [reply src/schema/sources.ts:1006]: I get it but it should still compare with `squad.memberPostingRank`, because if we introduce squads where only admins can post this would no longer work.

Below is more correct I would say (please check):

```suggestion
  if (memberRank >= sourceRoleRank.moderator && memberRank >= squad.memberPostingRank) {
  return true;  
}
```

**Review by capJavert — APPROVED** [review]

**Pull request merged**
